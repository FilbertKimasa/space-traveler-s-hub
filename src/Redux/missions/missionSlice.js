import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const apiUrl = 'https://api.spacexdata.com/v3/missions';
const initialState = {
  missions: [],
  isLoading: false,
  error: undefined,
};

export const fetchMission = createAsyncThunk('mission/fetchMission', async () => {
  const response = await axios.get(apiUrl);
  return response.data;
});

const missionSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {},

  extraReducers: (builder) => {
    builder
      .addCase(fetchMission.pending, (state) => {
        state.isLoading = true;
      })

      .addCase(fetchMission.fulfilled, (state, action) => {
        state.isLoading = false;
        state.missions = action.payload.map((mission) => ({
          id: mission.mission_id,
          mission_name: mission.mission_name,
          description: mission.description,

        }));
      })
      .addCase(fetchMission.rejected, (state, action) => {
        state.isLoading = false;
        state.error = action.error.message;
      });
  },
});

export default missionSlice.reducer;
