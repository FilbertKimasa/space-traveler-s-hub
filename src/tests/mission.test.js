import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import MissionList from '../components/mission';
import { fetchMission, joinMission } from '../redux/missions/missionSlice';
import '@testing-library/jest-dom';


// Create a mock store
const mockStore = configureStore([]);

describe('MissionList component', () => {
  let store;

  beforeEach(() => {
    // Initialize the store with initial state
    store = mockStore({
      missions: {
        missions: [
          {
            id: 1,
            mission_name: 'Mission 1',
            description: 'Description 1',
            join: false,
          },
          {
            id: 2,
            mission_name: 'Mission 2',
            description: 'Description 2',
            join: true,
          },
        ],
        isLoading: false,
        error: null,
      },
    });

    // Mock the dispatch function
    store.dispatch = jest.fn();
  });

  test('renders mission list', () => {
    render(
      <Provider store={store}>
        <MissionList />
      </Provider>
    );

    expect(screen.getByText('Mission 1')).toBeInTheDocument();
    expect(screen.getByText('Description 1')).toBeInTheDocument();
    expect(screen.getByText('Mission 2')).toBeInTheDocument();
    expect(screen.getByText('Description 2')).toBeInTheDocument();

    expect(screen.container).toMatchSnapshot();
  })
});