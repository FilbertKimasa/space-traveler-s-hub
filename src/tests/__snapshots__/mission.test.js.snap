// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissionList component dispatches joinMission action on button click 1`] = `Array []`;

exports[`MissionList component renders mission list 1`] = `undefined`;
